### ISSUE TITLE
[BUG] Date suggestions have flipped day/month

### ISSUE THREAD

**Issue by plu** [paperless-ngx/paperless-ngx#3647]
### Description

I have not set any `PAPERLESS_DATE_ORDER`, so I assume it's using the default `DMY`. This matches the date (the one on the right, 02.10.2020 aka 2nd of October 2020) for this document here:

![Screenshot 2023-06-21 at 08 35 55](https://github.com/paperless-ngx/paperless-ngx/assets/31597/45a94915-8441-460e-81e4-69d8d811416f)

In the web UI I have selected the ISO 8601 format. So I'd expect to see a suggestion `2020-10-02`.

Any idea why I'm getting `2020-02-10` instead?

### Steps to reproduce

Not sure.

### Webserver logs

```bash
Not sure this would help, but I'm happy to provide any other logs, if you point me in the right direction.
```

### Paperless-ngx version

1.16.1

### Host OS

Debian GNU/Linux 12

### Installation method

Docker - official image

### Browser

Safari Version 16.3 (18614.4.6.1.6)

**Comment by plu** [comment]
An easy solution would be to treat the big expression also smaller sub expressions. This would lead to possibly more found dates, not sure that is an issue?

```
>>> DATE_REGEX = (
...     re.compile(r"(\b|(?!=([_-])))([0-9]{1,2})[\.\/-]([0-9]{1,2})[\.\/-]([0-9]{4}|[0-9]{2})(\b|(?=([_-])))"),  # noqa: E501
...     re.compile(r"(\b|(?!=([_-])))([0-9]{4}|[0-9]{2})[\.\/-]([0-9]{1,2})[\.\/-]([0-9]{1,2})(\b|(?=([_-])))"),  # noqa: E501
...     re.compile(r"(\b|(?!=([_-])))([0-9]{1,2}[\. ]+[^ ]{3,9} ([0-9]{4}|[0-9]{2}))(\b|(?=([_-])))"),  # noqa: E501
...     re.compile(r"(\b|(?!=([_-])))([^\W\d_]{3,9} [0-9]{1,2}, ([0-9]{4}))(\b|(?=([_-])))"),
...     re.compile(r"(\b|(?!=([_-])))([^\W\d_]{3,9} [0-9]{4})(\b|(?=([_-])))"),
...     re.compile(r"(\b|(?!=([_-])))(\b[0-9]{1,2}[ \.\/-][A-Z]{3}[ \.\/-][0-9]{4})(\b|(?=([_-])))")  # noqa: E501
... )
>>>
>>> content = u'Finanzreport Nr. 9 per 01.10.2020 02.10.2020'
>>> for r in DATE_REGEX:
...     for m in re.finditer(r, content):
...             print(m.group(0))
...
01.10.2020
02.10.2020
9 per 01
10.2020 02
```

I have no development set up locally (on macOS, haven't properly tried getting it working yet).

Is there anybody who would be willing to try this change against the tests in [date_parsing.py](https://github.com/paperless-ngx/paperless-ngx/blob/dev/src/documents/tests/test_date_parsing.py), and possibly also add another test for `Finanzreport Nr. 9 per 01.10.2020 02.10.2020`? :)

**Comment by stumpylog** [comment]
The PAPERLESS_DATE_ORDER is only a suggestion to dateutil anyway.  And the expressions are just a best effort to find something datelike. Changing them is not going to be a small task without impacts everywhere

**Comment by plu** [comment]
Would you be fine with a PR that adds a new environment variable so I can inject a custom regular expression and superseed whatever is set in `DATE_REGEX`?

**Comment by stumpylog** [comment]
Your proposed change to the regex breaks this test: https://github.com/paperless-ngx/paperless-ngx/blob/dev/src/documents/tests/test_date_parsing.py#L164

It doesn't extract the July 22nd, 2022  date, instead getting July 1st, 2022

**Comment by plu** [comment]
Yep, let's forget about that change :). What about my other proposal, adding a new environment variable?

**Comment by stumpylog** [comment]
What about kind of "tuning" the regular expressions, based on the configured MDY ordering?  Some sort of thing so there's a set that's a little more likely to prefer datelike  strings in the ordering the user has configured, while still keeping the other matches as a backup?
